File: backend/app/schemas/team.py
# app/schemas/team.py
from typing import List, Optional
from pydantic import BaseModel, Field as PydanticField
from datetime import datetime

from app.models.team import TeamBase, TeamRole, TeamStatus, TeamVisibility
from app.models.user import UserPublic # Import UserPublic for member details

# --- Team Schemas ---

# Properties to receive via API on creation
class TeamCreate(TeamBase):
    name: str = PydanticField(..., max_length=50)
    tag: Optional[str] = PydanticField(None, max_length=10)


# Properties to receive via API on update
class TeamUpdate(BaseModel):
    name: Optional[str] = PydanticField(None, max_length=50)
    tag: Optional[str] = PydanticField(None, max_length=10)
    description: Optional[str] = PydanticField(None, max_length=500)
    logo_url: Optional[str] = None
    banner_url: Optional[str] = None
    status: Optional[TeamStatus] = None
    visibility: Optional[TeamVisibility] = None


# Properties shared by models stored in DB
class TeamInDBBase(TeamBase):
    id: int
    leader_id: int
    creation_date: datetime

    class Config:
        from_attributes = True # Pydantic V2 alias for orm_mode


# Properties to return to client (base team info)
class TeamRead(TeamInDBBase):
    pass


# --- Team Member Schemas ---

class TeamMemberBase(BaseModel):
    team_id: int
    user_id: int
    role: TeamRole


class TeamMemberRead(TeamMemberBase):
    join_date: datetime
    user: UserPublic # Include public user details

    class Config:
        from_attributes = True


# Properties to return team details including members
class TeamReadWithMembers(TeamRead):
    leader: UserPublic # Include leader details
    members: List[TeamMemberRead] = []


# Schema for transferring leadership
class TeamTransferLeadership(BaseModel):
    new_leader_user_id: int

# Schema for updating a member's role
class TeamMemberUpdateRole(BaseModel):
    role: TeamRole

# Schema for adding a member (simplified for now)
# In a real app, this might involve invites or requests
class TeamAddMember(BaseModel):
    user_id: int
    role: TeamRole = TeamRole.MEMBER # Default role when adding 

File: frontend/src/types/team.ts
import { UserPublic } from "./user"; // Assuming you have a user type file

export enum TeamRole {
    LEADER = "leader",
    OFFICER = "officer",
    MEMBER = "member",
    SUBSTITUTE = "substitute",
}

export enum TeamStatus {
    ACTIVE = "active",
    LOOKING_FOR_MEMBERS = "looking_for_members",
    DISBANDED = "disbanded",
    PRIVATE = "private",
}

export enum TeamVisibility {
    PUBLIC = "public",
    INVITE_ONLY = "invite_only",
}

// Base interface shared by create/update/read
export interface TeamBase {
    name: string;
    tag?: string | null;
    description?: string | null;
    logo_url?: string | null;
    banner_url?: string | null;
    status: TeamStatus;
    visibility: TeamVisibility;
}

// Interface for creating a team
export interface TeamCreate extends Omit<TeamBase, 'status' | 'visibility'> {
    // Status and visibility usually default on creation or are set explicitly
    name: string; // Make name required
    tag?: string | null;
}

// Interface for updating a team (all fields optional)
export interface TeamUpdate {
    name?: string | null;
    tag?: string | null;
    description?: string | null;
    logo_url?: string | null;
    banner_url?: string | null;
    status?: TeamStatus | null;
    visibility?: TeamVisibility | null;
}

// Interface for reading basic team data
export interface TeamRead extends TeamBase {
    id: number;
    leader_id: number;
    creation_date: string; // Dates are typically strings in JSON
}

// Interface for reading team member data
export interface TeamMemberRead {
    team_id: number;
    user_id: number;
    role: TeamRole;
    join_date: string;
    user: UserPublic; // Include public user details
}

// Interface for reading team data with leader and members included
export interface TeamReadWithMembers extends TeamRead {
    leader: UserPublic;
    members: TeamMemberRead[];
}

// Interface for adding a member (request body)
export interface TeamAddMemberPayload {
    user_id: number;
    role?: TeamRole; // Role might default on backend
}

// Interface for updating a member's role (request body)
export interface TeamMemberUpdateRolePayload {
    role: TeamRole;
}

// Interface for transferring leadership (request body)
export interface TeamTransferLeadershipPayload {
    new_leader_user_id: number;
} 